operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1728 @@
+					err := ipam.handleNewPool(clbi)
+					if err != nil {
+						syncErrors = append(syncErrors, err)
+					}
+
+					return nil
+				},
+				func(k resource.Key, clbi *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					syncErrors = append(syncErrors, errors.New("Got a Pool deleted event while syncing"))
+
+					return nil
+				},
+			)
+		}
+	}
+
+	// Then sync all services
+	svcSync := false
+	for !svcSync {
+		select {
+		case event := <-svcChan:
+			event.Handle(
+				func() error {
+					svcSync = true
+
+					err := ipam.satisfyServices()
+					if err != nil {
+						syncErrors = append(syncErrors, err)
+						return nil
+					}
+
+					err = ipam.updateAllPoolCounts()
+					if err != nil {
+						syncErrors = append(syncErrors, err)
+						return nil
+					}
+
+					return nil
+				},
+				func(k resource.Key, svc *core_v1.Service) error {
+					err := ipam.handleUpsertService(svc)
+					if err != nil {
+						syncErrors = append(syncErrors, err)
+					}
+
+					return nil
+				},
+				func(k resource.Key, svc *core_v1.Service) error {
+					syncErrors = append(syncErrors, errors.New("Got a Pool deleted event while syncing"))
+
+					return nil
+				},
+			)
+		}
+	}
+
+	if len(syncErrors) > 0 {
+		return multierr.Combine(syncErrors...)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) poolOnUpsert(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	// Deep copy so we get a version we are allowed to update
+	pool = pool.DeepCopy()
+
+	var err error
+	if _, exists := ipam.poolStore.GetByName(pool.GetName()); exists {
+		ipam.logger.Debugf("Updating pool '%s'", pool.GetName())
+		err = ipam.handlePoolModified(pool)
+		if err != nil {
+			ipam.logger.WithError(err).Error("Error in handlePoolModified()")
+		}
+	} else {
+		ipam.logger.Debugf("Adding pool '%s'", pool.GetName())
+		err = ipam.handleNewPool(pool)
+		if err != nil {
+			ipam.logger.WithError(err).Error("Error in handleNewPool()")
+		}
+	}
+	if err != nil {
+		if isRetryableError(err) {
+			// Return the error, this will trigger the service resource to retry this service at a later time in the hope
+			// that any transient error is resolved by then.
+			return err
+		}
+
+		return nil
+	}
+
+	err = ipam.settleConflicts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in settleConflicts()")
+
+		if isRetryableError(err) {
+			// Return the error, this will trigger the service resource to retry this service at a later time in the hope
+			// that any transient error is resolved by then.
+			return err
+		}
+
+		return nil
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+
+		if isRetryableError(err) {
+			// Return the error, this will trigger the service resource to retry this service at a later time in the hope
+			// that any transient error is resolved by then.
+			return err
+		}
+
+		return nil
```

Comment:
Why are we ignoring "non-retryable" errors? What even is a "non-retriable" error? Why don't we just always retry?

I would really just always retry no matter what goes wrong and allow it to bubble up through the `errChan` in `Run()` and then there decide to either restart or crash, or have infinite retries with exponential back-off. Basically build this so that you never give up,  but rate limit so you don't try too hard. Trying to decide what is and isn't retryable and ignoring errors seems dangerous.

Consider returning a wrapped error with more context and moving the error logging into the retry handler to avoid all the boilerplate that repeats here.

Reply:
All errors now have infinite retries with exponential backoff to a max of 5m, see the operator/cmd/resources.go. That should do it.